discovery-handlers/onvif/src/discovery_handler.rs

```diff
@@ -72,42 +72,51 @@ impl DiscoveryHandler for DiscoveryHandlerImpl {
         let discovery_handler_config: OnvifDiscoveryDetails =
             deserialize_discovery_details(&discover_request.discovery_details)
                 .map_err(|e| tonic::Status::new(tonic::Code::InvalidArgument, format!("{}", e)))?;
-        let mut cameras: Vec<Device> = Vec::new();
         tokio::spawn(async move {
+            let mut previous_cameras = Vec::new();
+            let mut filtered_camera_devices = HashMap::new();
             loop {
+                let mut changed_camera_list = false;
                 let onvif_query = OnvifQueryImpl {};
 
                 trace!("discover - filters:{:?}", &discovery_handler_config,);
-                let discovered_onvif_cameras = util::simple_onvif_discover(Duration::from_secs(
-                    discovery_handler_config.discovery_timeout_seconds as u64,
-                ))
-                .await
-                .unwrap();
-                trace!("discover - discovered:{:?}", &discovered_onvif_cameras,);
-                // apply_filters never returns an error -- safe to unwrap
-                let filtered_onvif_cameras = apply_filters(
-                    &discovery_handler_config,
-                    discovered_onvif_cameras,
-                    &onvif_query,
+                let mut socket = util::get_discovery_response_socket().await.unwrap();
+                let latest_cameras = util::simple_onvif_discover(
+                    &mut socket,
+                    discovery_handler_config.scopes.as_ref(),
+                    Duration::from_secs(discovery_handler_config.discovery_timeout_seconds as u64),
                 )
                 .await
                 .unwrap();
-                trace!("discover - filtered:{:?}", &filtered_onvif_cameras);
-                let mut changed_camera_list = false;
-                let mut matching_camera_count = 0;
-                filtered_onvif_cameras.iter().for_each(|camera| {
-                    if !cameras.contains(camera) {
+                trace!("discover - discovered:{:?}", &latest_cameras);
+                // Remove cameras that have gone offline
+                previous_cameras.iter().for_each(|c| {
```

Comment:
The retry timeout is effectively 10 seconds since that is the discovery interval. If they were to come back online on the next discovery check they will once again be reported to the agent

Reply:
So if on second X camera A is present, on second X+9 it disappears, on X+10 another scan does not find it, so it removes it from Akri instance, on X+11 camera comes back (so camera had intermittent connectivity for 2 seconds), we still end up evicting pods that were using the camera or is my understanding wrong?

Reply:
Your scenario does/would not occur. The discovery handler logic is different from the agent logic. The Discovery Handler reports to the Agent whenever there are changes in device visibility, checking for changes every 10 seconds. The Agent then decides what to do with that information. Currently, for shared devices, it gives a 5 minute timeout before deleting instances, triggering the controller to bring down pods.

Reply:
> Does this logic need some retry timeout before removing the cameras or is that handled elsewhere?

To summarize: the removing of instances and brokers is handled elsewhere, in the Agent and Controller respectively after a 5 minute period of the Discovery Handler not reporting the device.